refactor: Get rid of more redundant chain methods

ryanofsky opened the pull request:
This just drops three interfaces::Chain methods replacing them with other calls.

Motivation for removing these chain methods:

- Need to get rid of findFirstBlockWithTimeAndHeight for #10102, which doesn't support overloaded methods
- Followup from https://github.com/bitcoin/bitcoin/pull/16426#discussion_r412487403
- phantomcircuit comments about findNextBlock test http://www.erisian.com.au/bitcoin-core-dev/log-2020-06-06.html#l-214

Behavior is not changing in any way here. A TODO comment in ScanForWalletTransactions was removed, but just because it was invalid (see https://github.com/bitcoin/bitcoin/pull/19195#discussion_r448020762), not because it was implemented.

ryanofsky committed:
refactor: Get rid of more redundant chain methods

This just drops three interfaces::Chain methods replacing them with other calls.

Motivation for removing these chain methods:

- Need to get rid of findFirstBlockWithTimeAndHeight for #10102, which doesn't
  support overloaded methods
- Followup from
  https://github.com/bitcoin/bitcoin/pull/16426#discussion_r412487403
- phantomcircuit comments about findNextBlock test
  http://www.erisian.com.au/bitcoin-core-dev/log-2020-06-06.html#l-214

Behavior is not changing in any way here. A TODO comment in
ScanForWalletTransactions was removed, but just because it was invalid (see
https://github.com/bitcoin/bitcoin/pull/19195#discussion_r448020762), not
because it was implemented.

ryanofsky commented:
Updated 0c9dea569a41c1fa50b2409fabc531cfa663ad51 -> 87bc7618cf561ed97835558b6e0c19affdc2aaeb ([`pr/next.1`](https://github.com/ryanofsky/bitcoin/commits/pr/next.1) -> [`pr/next.2`](https://github.com/ryanofsky/bitcoin/commits/pr/next.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/next.1..pr/next.2)) adding comments and removing stale TODO. Still no changes in behavior

promag reviewed (commented):
Concept ACK.

laanwj commented:
Concept ACK. I find this code really hard to review.

promag commented on src/interfaces/chain.cpp:
```diff
@@ -46,6 +46,8 @@ bool FillBlock(const CBlockIndex* index, const FoundBlock& block, UniqueLock<Rec
+    if (block.m_next_block) FillBlock(ChainActive()[index->nHeight] == index ? ChainActive()[index->nHeight + 1] : nullptr, *block.m_next_block, lock);
     if (block.m_data) {
         REVERSE_LOCK(lock);
         if (!ReadBlockFromDisk(*block.m_data, index, Params().GetConsensus())) block.m_data->SetNull();
```
Unrelated, but why return `true` if `ReadBlockfromDisk` fails?

MarcoFalke commented on src/interfaces/chain.cpp:
```diff
@@ -46,6 +46,8 @@ bool FillBlock(const CBlockIndex* index, const FoundBlock& block, UniqueLock<Rec
     if (block.m_max_time) *block.m_max_time = index->GetBlockTimeMax();
     if (block.m_mtp_time) *block.m_mtp_time = index->GetMedianTimePast();
+    if (block.m_in_active_chain) *block.m_in_active_chain = ::ChainActive()[index->nHeight] == index;
+    if (block.m_next_block) FillBlock(ChainActive()[index->nHeight] == index ? ChainActive()[index->nHeight + 1] : nullptr, *block.m_next_block, lock);
```
Also, new code should probably use node->chainman.activeChain

ryanofsky committed:
refactor: Replace uses ChainActive() in interfaces/chain.cpp

Suggested https://github.com/bitcoin/bitcoin/pull/19425#discussion_r456236407

ryanofsky reviewed (commented):
Updated 91ac0388e88f69531bd7c957e9d0e5ee5e36b347 -> 7ed165a586ac7a32f4f80be0ead81bee0250f430 ([`pr/next.3`](https://github.com/ryanofsky/bitcoin/commits/pr/next.3) -> [`pr/next.4`](https://github.com/ryanofsky/bitcoin/commits/pr/next.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/next.3..pr/next.4)) adding commit to get rid of ChainActive

MarcoFalke commented on src/interfaces/chain.cpp:
```diff
@@ -358,7 +344,8 @@ class ChainImpl : public Chain
         if (!old_tip.IsNull()) {
             LOCK(::cs_main);
-            if (old_tip == ::ChainActive().Tip()->GetBlockHash()) return;
+            const CChain& active = m_node.chainman->ActiveChain();
```
```suggestion
            const CChain& active = Assert(m_node.chainman)->ActiveChain();
```

style-nit, no strong opinion, though.

ryanofsky reviewed (commented):
Rebased 7ed165a586ac7a32f4f80be0ead81bee0250f430 -> 6e10c2b8832d24cc010fd39b23e23543e00012e5 ([`pr/next.4`](https://github.com/ryanofsky/bitcoin/commits/pr/next.4) -> [`pr/next.5`](https://github.com/ryanofsky/bitcoin/commits/pr/next.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/next.4-rebase..pr/next.5)) due to conflict with #19098

laanwj commented:
I managed to review the code and am fairly sure it is correct.
Code review ACK 6e10c2b8832d24cc010fd39b23e23543e00012e5

ryanofsky commented:
Rebased 6e10c2b8832d24cc010fd39b23e23543e00012e5 -> 9931714c0a1347d377b73f50e900c1138081d1c2 ([`pr/next.5`](https://github.com/ryanofsky/bitcoin/commits/pr/next.5) -> [`pr/next.6`](https://github.com/ryanofsky/bitcoin/commits/pr/next.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/next.5-rebase..pr/next.6)) due to conflict with #19619

dongcarl commented:
Code Review ACK 9931714c0a1347d377b73f50e900c1138081d1c2

- 97e683d1a1f57dd8fac080eec1a0fe83419df7ac doesn't seem to change behaviour, and seems to match FoundBlock's conventions. I found the calling conventions to be a bit weird but they're still understandable. Perhaps someone can tell me why we do it this way, is it for efficiency?
- 9931714c0a1347d377b73f50e900c1138081d1c2 arrived at the same conclusion as I did for resolving multiple `NodeContext`s. See: [`ef1b08c` (#20158)](https://github.com/bitcoin/bitcoin/pull/20158/commits/ef1b08c0738da2248fbddc0190320f5ed34b7f44) and [`a9539e7` (#20158)](https://github.com/bitcoin/bitcoin/pull/20158/commits/a9539e725bb109fb8d98970223fa8eff6a582b4a). Elimination of `::ChainActive()` also makes my diff for #20158 shorter.

ryanofsky commented:
Thanks for reviewing!

> Perhaps someone can tell me why we do it this way, is it for efficiency?

It's called a "fluent interface" and it's just a way emulating a way of named parameters in languages which don't support them: https://en.wikipedia.org/wiki/Fluent_interface. The alternatives are adding unnamed parameters or adding method overloads

MarcoFalke commented on src/bench/wallet_balance.cpp:
```diff
@@ -24,15 +24,14 @@ static void WalletBalance(benchmark::Bench& bench, const bool set_dirty, const b
-    NodeContext node;
-    std::unique_ptr<interfaces::Chain> chain = interfaces::MakeChain(node);
-    CWallet wallet{chain.get(), "", CreateMockWalletDatabase()};
+    test_setup.m_node.chain = interfaces::MakeChain(test_setup.m_node);
```
Can you explain what this is doing and why the redundant call is needed?


```diff
diff --git a/src/bench/wallet_balance.cpp b/src/bench/wallet_balance.cpp
index 98eac80055..aa436ee3ea 100644
--- a/src/bench/wallet_balance.cpp
+++ b/src/bench/wallet_balance.cpp
@@ -24,7 +24,6 @@ static void WalletBalance(benchmark::Bench& bench, const bool set_dirty, const b
 
     const auto& ADDRESS_WATCHONLY = ADDRESS_BCRT1_UNSPENDABLE;
 
-    test_setup.m_node.chain = interfaces::MakeChain(test_setup.m_node);
     CWallet wallet{test_setup.m_node.chain.get(), "", CreateMockWalletDatabase()};
     {
         wallet.SetupLegacyScriptPubKeyMan();
diff --git a/src/test/interfaces_tests.cpp b/src/test/interfaces_tests.cpp
index b8c4bc8605..73463b071e 100644
--- a/src/test/interfaces_tests.cpp
+++ b/src/test/interfaces_tests.cpp
@@ -17,7 +17,7 @@ BOOST_FIXTURE_TEST_SUITE(interfaces_tests, TestChain100Setup)
 
 BOOST_AUTO_TEST_CASE(findBlock)
 {
-    auto chain = interfaces::MakeChain(m_node);
+    auto& chain = m_node.chain;
     const CChain& active = Assert(m_node.chainman)->ActiveChain();
 
     uint256 hash;
@@ -61,7 +61,7 @@ BOOST_AUTO_TEST_CASE(findBlock)
 
 BOOST_AUTO_TEST_CASE(findFirstBlockWithTimeAndHeight)
 {
-    auto chain = interfaces::MakeChain(m_node);
+    auto& chain = m_node.chain;
     const CChain& active = Assert(m_node.chainman)->ActiveChain();
     uint256 hash;
     int height;
@@ -73,7 +73,7 @@ BOOST_AUTO_TEST_CASE(findFirstBlockWithTimeAndHeight)
 
 BOOST_AUTO_TEST_CASE(findAncestorByHeight)
 {
-    auto chain = interfaces::MakeChain(m_node);
+    auto& chain = m_node.chain;
     const CChain& active = Assert(m_node.chainman)->ActiveChain();
     uint256 hash;
     BOOST_CHECK(chain->findAncestorByHeight(active[20]->GetBlockHash(), 10, FoundBlock().hash(hash)));
@@ -83,7 +83,7 @@ BOOST_AUTO_TEST_CASE(findAncestorByHeight)
 
 BOOST_AUTO_TEST_CASE(findAncestorByHash)
 {
-    auto chain = interfaces::MakeChain(m_node);
+    auto& chain = m_node.chain;
     const CChain& active = Assert(m_node.chainman)->ActiveChain();
     int height = -1;
     BOOST_CHECK(chain->findAncestorByHash(active[20]->GetBlockHash(), active[10]->GetBlockHash(), FoundBlock().height(height)));
@@ -93,7 +93,7 @@ BOOST_AUTO_TEST_CASE(findAncestorByHash)
 
 BOOST_AUTO_TEST_CASE(findCommonAncestor)
 {
-    auto chain = interfaces::MakeChain(m_node);
+    auto& chain = m_node.chain;
     const CChain& active = Assert(m_node.chainman)->ActiveChain();
     auto* orig_tip = active.Tip();
     for (int i = 0; i < 10; ++i) {
@@ -123,7 +123,7 @@ BOOST_AUTO_TEST_CASE(findCommonAncestor)
 
 BOOST_AUTO_TEST_CASE(hasBlocks)
 {
-    auto chain = interfaces::MakeChain(m_node);
+    auto& chain = m_node.chain;
     const CChain& active = Assert(m_node.chainman)->ActiveChain();
 
     // Test ranges
diff --git a/src/wallet/test/coinselector_tests.cpp b/src/wallet/test/coinselector_tests.cpp
index f38ccba384..4127cd45f8 100644
--- a/src/wallet/test/coinselector_tests.cpp
+++ b/src/wallet/test/coinselector_tests.cpp
@@ -283,7 +283,7 @@ BOOST_AUTO_TEST_CASE(bnb_search_test)
     // Make sure that can use BnB when there are preset inputs
     empty_wallet();
     {
-        std::unique_ptr<CWallet> wallet = MakeUnique<CWallet>(m_chain.get(), "", CreateMockWalletDatabase());
+        std::unique_ptr<CWallet> wallet = MakeUnique<CWallet>(m_node.chain.get(), "", CreateMockWalletDatabase());
         bool firstRun;
         wallet->LoadWallet(firstRun);
         wallet->SetupLegacyScriptPubKeyMan();
diff --git a/src/wallet/test/init_test_fixture.cpp b/src/wallet/test/init_test_fixture.cpp
index c80310045a..334e4ae0d8 100644
--- a/src/wallet/test/init_test_fixture.cpp
+++ b/src/wallet/test/init_test_fixture.cpp
@@ -10,7 +10,7 @@
 
 InitWalletDirTestingSetup::InitWalletDirTestingSetup(const std::string& chainName) : BasicTestingSetup(chainName)
 {
-    m_wallet_client = MakeWalletClient(*m_chain, *Assert(m_node.args));
+    m_wallet_client = MakeWalletClient(*m_node.chain, *Assert(m_node.args));
 
     std::string sep;
     sep += fs::path::preferred_separator;
diff --git a/src/wallet/test/init_test_fixture.h b/src/wallet/test/init_test_fixture.h
index f5bade77df..f666c45a34 100644
--- a/src/wallet/test/init_test_fixture.h
+++ b/src/wallet/test/init_test_fixture.h
@@ -19,7 +19,6 @@ struct InitWalletDirTestingSetup: public BasicTestingSetup {
     fs::path m_datadir;
     fs::path m_cwd;
     std::map<std::string, fs::path> m_walletdir_path_cases;
-    std::unique_ptr<interfaces::Chain> m_chain = interfaces::MakeChain(m_node);
     std::unique_ptr<interfaces::WalletClient> m_wallet_client;
 };
 
diff --git a/src/wallet/test/ismine_tests.cpp b/src/wallet/test/ismine_tests.cpp
index d5aed99d99..0ef8b9c4bf 100644
--- a/src/wallet/test/ismine_tests.cpp
+++ b/src/wallet/test/ismine_tests.cpp
@@ -27,8 +27,7 @@ BOOST_AUTO_TEST_CASE(ismine_standard)
     CKey uncompressedKey;
     uncompressedKey.MakeNewKey(false);
     CPubKey uncompressedPubkey = uncompressedKey.GetPubKey();
-    NodeContext node;
-    std::unique_ptr<interfaces::Chain> chain = interfaces::MakeChain(node);
+    std::unique_ptr<interfaces::Chain>& chain = m_node.chain;
 
     CScript scriptPubKey;
     isminetype result;
diff --git a/src/wallet/test/scriptpubkeyman_tests.cpp b/src/wallet/test/scriptpubkeyman_tests.cpp
index f7c1337b0d..347a436429 100644
--- a/src/wallet/test/scriptpubkeyman_tests.cpp
+++ b/src/wallet/test/scriptpubkeyman_tests.cpp
@@ -17,9 +17,7 @@ BOOST_FIXTURE_TEST_SUITE(scriptpubkeyman_tests, BasicTestingSetup)
 BOOST_AUTO_TEST_CASE(CanProvide)
 {
     // Set up wallet and keyman variables.
-    NodeContext node;
-    std::unique_ptr<interfaces::Chain> chain = interfaces::MakeChain(node);
-    CWallet wallet(chain.get(), "", CreateDummyWalletDatabase());
+    CWallet wallet(m_node.chain.get(), "", CreateDummyWalletDatabase());
     LegacyScriptPubKeyMan& keyman = *wallet.GetOrCreateLegacyScriptPubKeyMan();
 
     // Make a 1 of 2 multisig script
diff --git a/src/wallet/test/wallet_test_fixture.cpp b/src/wallet/test/wallet_test_fixture.cpp
index 4d6f427618..badf2eb459 100644
--- a/src/wallet/test/wallet_test_fixture.cpp
+++ b/src/wallet/test/wallet_test_fixture.cpp
@@ -6,10 +6,10 @@
 
 WalletTestingSetup::WalletTestingSetup(const std::string& chainName)
     : TestingSetup(chainName),
-      m_wallet(m_chain.get(), "", CreateMockWalletDatabase())
+      m_wallet(m_node.chain.get(), "", CreateMockWalletDatabase())
 {
     bool fFirstRun;
     m_wallet.LoadWallet(fFirstRun);
-    m_chain_notifications_handler = m_chain->handleNotifications({ &m_wallet, [](CWallet*) {} });
+    m_chain_notifications_handler = m_node.chain->handleNotifications({ &m_wallet, [](CWallet*) {} });
     m_wallet_client->registerRpcs();
 }
diff --git a/src/wallet/test/wallet_test_fixture.h b/src/wallet/test/wallet_test_fixture.h
index ba8a5ff1f3..ab7fb8c42b 100644
--- a/src/wallet/test/wallet_test_fixture.h
+++ b/src/wallet/test/wallet_test_fixture.h
@@ -20,8 +20,7 @@
 struct WalletTestingSetup : public TestingSetup {
     explicit WalletTestingSetup(const std::string& chainName = CBaseChainParams::MAIN);
 
-    std::unique_ptr<interfaces::Chain> m_chain = interfaces::MakeChain(m_node);
-    std::unique_ptr<interfaces::WalletClient> m_wallet_client = interfaces::MakeWalletClient(*m_chain, *Assert(m_node.args));
+    std::unique_ptr<interfaces::WalletClient> m_wallet_client = interfaces::MakeWalletClient(*m_node.chain, *Assert(m_node.args));
     CWallet m_wallet;
     std::unique_ptr<interfaces::Handler> m_chain_notifications_handler;
 };
diff --git a/src/wallet/test/wallet_tests.cpp b/src/wallet/test/wallet_tests.cpp
index 156be053d0..eb0bbb6ccc 100644
--- a/src/wallet/test/wallet_tests.cpp
+++ b/src/wallet/test/wallet_tests.cpp
@@ -83,8 +83,6 @@ BOOST_FIXTURE_TEST_CASE(scan_for_wallet_transactions, TestChain100Setup)
     CreateAndProcessBlock({}, GetScriptForRawPubKey(coinbaseKey.GetPubKey()));
     CBlockIndex* newTip = ::ChainActive().Tip();
 
-    m_node.chain = interfaces::MakeChain(m_node);
-
     // Verify ScanForWalletTransactions fails to read an unknown start block.
     {
         CWallet wallet(m_node.chain.get(), "", CreateDummyWalletDatabase());
@@ -182,8 +180,6 @@ BOOST_FIXTURE_TEST_CASE(importmulti_rescan, TestChain100Setup)
     CreateAndProcessBlock({}, GetScriptForRawPubKey(coinbaseKey.GetPubKey()));
     CBlockIndex* newTip = ::ChainActive().Tip();
 
-    m_node.chain = interfaces::MakeChain(m_node);
-
     // Prune the older block file.
     {
         LOCK(cs_main);
@@ -253,8 +249,6 @@ BOOST_FIXTURE_TEST_CASE(importwallet_rescan, TestChain100Setup)
     SetMockTime(KEY_TIME);
     m_coinbase_txns.emplace_back(CreateAndProcessBlock({}, GetScriptForRawPubKey(coinbaseKey.GetPubKey())).vtx[0]);
 
-    m_node.chain = interfaces::MakeChain(m_node);
-
     std::string backup_file = (GetDataDir() / "wallet.backup").string();
 
     // Import key into wallet and call dumpwallet to create backup file.
@@ -489,7 +483,6 @@ public:
     ListCoinsTestingSetup()
     {
         CreateAndProcessBlock({}, GetScriptForRawPubKey(coinbaseKey.GetPubKey()));
-        m_node.chain = interfaces::MakeChain(m_node);
         wallet = MakeUnique<CWallet>(m_node.chain.get(), "", CreateMockWalletDatabase());
         {
             LOCK2(wallet->cs_wallet, ::cs_main);
@@ -701,7 +694,6 @@ BOOST_FIXTURE_TEST_CASE(wallet_descriptor_test, BasicTestingSetup)
 BOOST_FIXTURE_TEST_CASE(CreateWallet, TestChain100Setup)
 {
     // Create new wallet with known key and unload it.
-    m_node.chain = interfaces::MakeChain(m_node);
     auto wallet = TestLoadWallet(*m_node.chain);
     CKey key;
     key.MakeNewKey(true);
@@ -794,8 +786,7 @@ BOOST_FIXTURE_TEST_CASE(CreateWallet, TestChain100Setup)
 
 BOOST_FIXTURE_TEST_CASE(ZapSelectTx, TestChain100Setup)
 {
-    auto chain = interfaces::MakeChain(m_node);
-    auto wallet = TestLoadWallet(*chain);
+    auto wallet = TestLoadWallet(*m_node.chain);
     CKey key;
     key.MakeNewKey(true);
     AddKey(*wallet, key);

dongcarl replied:
Right, looks like anything that has a fixture which inherits from `BasicTestingSetup` will have `m_node.chain` properly initialized: https://github.com/bitcoin/bitcoin/blob/a0489f3472f3799dc1ece32a59556fd239c4c14b/src/test/util/setup_common.cpp#L109

maflcko reviewed (approved):
cr ACK 9931714c0a1347d377b73f50e900c1138081d1c2 🍲

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

cr ACK 9931714c0a1347d377b73f50e900c1138081d1c2 🍲
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUhUOQwAomF3ZBN9ZKEQ+0iRMfzrYW8WoUmJUUGE+3exY2nV+2nnySpCT2N+nGg6
YnQk5UWMExmUSqK9JAD783Zfun1YLRASpfL7uZicLs3ZLMVM/4cFLnaoEk9RvrZk
zZdfdwRcPhhz0WLMjRCgFbubwWNxd2BqVObzUpfhenQBUlggrRn2auHsvJTa0vTN
19ZWvNDIgq9UYJVwJ47b2X4eu5bJeLcmKWuc+pGxchmAnYfAA6iDklIuG+3AVBWm
DdDNUttzy5z1o8Bd7s9m8b/EzrhiVUPRx3YsX4hWPCD+YlEb5ATLzERoiFYlqUTY
RJCuOZPRmLQRhqjRVdAklDVlz7BClOT4cMBwdZeBR0s8keEJqH+F8aCn0U72eFWq
ix8Fer2Fsj0YOvYIClryzamuFgPvA6e9D+idKdDjY6XXdoLsRljzRSejdOCcbVXu
wAdcVkI891WiN/EgBmWO9LHi+UCzwbQe1iuJ3j28fzKVcdrZGf6PBsGd+s1WS0XU
0UUQV8km
=T1NN
-----END PGP SIGNATURE-----
```

python: error while loading shared libraries: libpython3.8.so.1.0: cannot open shared object file: No such file or directory

</details>

dongcarl commented on src/interfaces/chain.h:
```diff
@@ -44,6 +44,10 @@ class FoundBlock
     FoundBlock& mtpTime(int64_t& mtp_time) { m_mtp_time = &mtp_time; return *this; }
+    //! Return whether block is in the active (most-work) chain.
+    FoundBlock& inActiveChain(bool& in_active_chain) { m_in_active_chain = &in_active_chain; return *this; }
+    //! Return next block in the active chain if current block is in the active chain.
```
Curious: Why does the passed bool reference have to be initialized to false? Wouldn't it be okay for it to be initialized to true as well?

MarcoFalke replied:
That would imply if the block isn't found it is assumed to be in the active chain? Seems a footgun just as dangerous

ryanofsky committed:
test: Remove no longer needed MakeChain calls

These calls are no longer needed after edc316020e8270dafc5e31465d532baebdafd3dd
from #19098 which started instantiating BasicTestingSetup.m_node.chain

Patch from MarcoFalke <falke.marco@gmail.com> in
https://github.com/bitcoin/bitcoin/pull/19425#discussion_r526701954

Co-authored-by: MarcoFalke <falke.marco@gmail.com>

ryanofsky reviewed (commented):
Rebased 9931714c0a1347d377b73f50e900c1138081d1c2 -> 5baa88fd38c8efa0e361636bb2c60af89d27b5d5 ([`pr/next.6`](https://github.com/ryanofsky/bitcoin/commits/pr/next.6) -> [`pr/next.7`](https://github.com/ryanofsky/bitcoin/commits/pr/next.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/next.6-rebase..pr/next.7)) due to conflict with #20494 and removed MakeChain calls no longer needed after #19098 as suggested

MarcoFalke commented:
re-ACK 5baa88fd38c8efa0e361636bb2c60af89d27b5d5 🕶

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

re-ACK 5baa88fd38c8efa0e361636bb2c60af89d27b5d5 🕶
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgRQAv6A0yyCNiZr7DYQtS7ubtoC310OFwRL/ghzzgjyS605iRdOHFWzvQxFLaP
9bKMhe5T5XMu7SE+0Ow3eXiVqXhCAsGqEXvdFxmPSqbmzCoG2bokdG6ZNNzsZ65/
IvzFdHkmLkT9Cj5knqZhZu9Qc24ncyw212cEQ674OQA53y+76HcE7cyJsS2kx56n
eBDajcd5wa9GaOSrT0SF4UOgotXCHX+NQgM+QSJRN/ZNAFUbGu4Q+/Qj9YMlvFvg
zG3TOSI8bi4PX/ygR/J8Ju4Sc4fosMWwsIwaLV0e6ErNEF3dtK/JVCzRvnjyQ38G
bId7rr1OBLi+pfQBB4oTPYDMgbf06R3cAV+jJ5rlAEcTWw+xMRfxhZEdZqdFGHOz
ZR1YuPn5EEibs2VmmQMqeuHmBVxYJzaWAjH9OwQI7qkE+CQAa6I/nmoEBmiRk7m4
z3YbBAY7XfaNH6l455wPcFhaxAcfSWvADHt24amYpmSSH0kvpaWVu9g4QzeCMqtY
yy+lKj94
=j9ux
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `19b54d707dfb2ae462bb723d421d5753739d3d2543051b56a0b9445ec9485431  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010819b54d707dfb2ae462bb723d421d5753739d3d2543051b56a0b9445ec9485431f010cd4fda73067f46652b723e5fe74cb8ab08fff0108693ae1c83d40d662f66062f5e31239608f1045fcf669bf0080b7df3494d69ec180083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01089d16c7f575d155069077b4d138d4f0808f1045fcf669cf008b4f55e6dbc85d4a80083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff010cec55446366129e66983ef3c11db011008f1045fcf669cf008ed28ab6f4232dcbf0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010ecfe180a3a6ce4de4035c62ca66e041908f1045fcf669cf008a6bffa91e10a68bd0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6d)

</details>

dongcarl commented:
ACK 5baa88fd38c8efa0e361636bb2c60af89d27b5d5
